linked collection creation with firebase user; made collection creati…

gnoahp opened the pull request:
…on go through POST api/services-lists

gnoahp commented:
Some thoughts I had while doing this:

Airtable  Changes (not sure if they have to be duplicated somewhere to push to prod):
added "userId", "Created Time", and "Last Updated" fields to "Resource Lists" table

1. do we really want to keep track of the user <-> list connection in both firebase and airtable/it feels like we're asking for trouble if they get out of sync (is just having the firebase uid in airtable enough?)

2. can we replace the "Created At" field in "Resource Lists" with the "Created Time" field (the "Created At" field is just a normal date and so does not automatically update when you create a new list/not sure if this is desired behavior)

notnazsty commented:
> Airtable Changes (not sure if they have to be duplicated somewhere to push to prod): added "userId", "Created Time", and "Last Updated" fields to "Resource Lists" table
> 
> Some thoughts I had while doing this:
> 
> 1. do we really want to keep track of the user <-> list connection in both firebase and airtable/it feels like we're asking for trouble if they get out of sync (is just having the firebase uid in airtable enough?)
> 2. can we replace the "Created At" field in "Resource Lists" with the "Created Time" field (the "Created At" field is just a normal date and so does not automatically update when you create a new list/not sure if this is desired behavior)

I was thinking that we could sync up the airtable data every so often to firebase and just pull everything from firebase, that way we don't need to do the caching or worry about airtable rate limits as much. But that's definitely something for the future

notnazsty commented:
This looks solid though, I don't see any issues

dangalan commented:
> Airtable Changes (not sure if they have to be duplicated somewhere to push to prod): added "userId", "Created Time", and "Last Updated" fields to "Resource Lists" table
> 
> Some thoughts I had while doing this:
> 
> 1. do we really want to keep track of the user <-> list connection in both firebase and airtable/it feels like we're asking for trouble if they get out of sync (is just having the firebase uid in airtable enough?)
> 2. can we replace the "Created At" field in "Resource Lists" with the "Created Time" field (the "Created At" field is just a normal date and so does not automatically update when you create a new list/not sure if this is desired behavior)

[Here is the prod Airtable](https://airtable.com/appF8rJZ3XAkzWOO9) if you wanna duplicate your changes there - you should have permission.

Could you camelCase the new time fields you added (e.g. `Created Time` -> `createdTime`)? Feel like the format makes more sense with code.

1. Good call out - I think we should _just_  have the Airtable collection ids on the Firebase so we don't need `userId` on the `Resource Lists` table.
2. Good catch here too - I thought when I made the `Created At` field it was the created time, not just a normal date. I deleted the field from Airtable if you could delete it from the model code too.

dangalan commented on hooks/useCreateCollection.ts:
```diff
@@ -61,25 +62,41 @@ export const useCreateCollection = (): useCreateCollectionReturn => {
-        resources: data.resources.map((resource) => resource.id),
-      },
-    ])
+    // This should be handled somewhere else (i.e. shouldn't be able to submit unless there's userInfo)
+    if (userInfo.user === null) {
+      createCollectionMutation([
+        {
+          // Set status to "Draft" b/c research team has to approve new collections
+          status: 'Draft',
+          name: data.name,
+          description: data.description,
+          creator: data.creator,
+          resources: data.resources.map((resource) => resource.id),
+        },
+      ])
+    } else {
+      createCollectionMutation([
+        {
+          // Set status to "Draft" b/c research team has to approve new collections
+          status: 'Draft',
+          name: data.name,
+          description: data.description,
+          creator: data.creator,
+          resources: data.resources.map((resource) => resource.id),
+          userId: userInfo.user.uid,
+        },
+      ])
+    }
```
Not sure if this is the best way to do but my plan was to have the UI prompt the user to login before they could submit so this endpoint would never be hit without a user being logged in.

A better way to do it could be to just submit `userInfo.user.uid` regardless of whether or not it's defined and check it in the endpoint and maybe throw an error there. Either way, we can get rid of this conditional.

dangalan replied:
The most like correct way to do this would prob be to add some Firebase auth token from when users are logged in to the request headers and check that in the backend but we can prob save that for a future MR.

dangalan commented on hooks/useCreateCollection.ts:
```diff
@@ -61,25 +62,41 @@ export const useCreateCollection = (): useCreateCollectionReturn => {
     useMutation(
       (createServicesListRequests: CreateServicesListRequest[]) =>
-        createServicesLists(createServicesListRequests),
+        postServicesList(createServicesListRequests),
```
Good catch !

dangalan reviewed (changes requested):
Some comments on the conditional you left a comment on in `useCreateCollection()` !

gnoahp commented:
1. got rid of userId field from airtable
2. changed field names to lastUpdated and createdTime;
3. simplified the conditional in useCreateCollection
4. replaced references to createdAt with createdTime
5. made user collections display most recently created at top (was displaying them in ascending order before)

dangalan reviewed (approved):
LGTM 🚀
